TERMINAL: Add RAM usage and detailed file size to `ls -l` output

HansLuft778 opened the pull request:
closes #1958 

I have updated the `ls -l` command to show RAM usage and byte-accurate file sizes. The file data is then formatted and passed to a `LongListItem` React component to render the table-like `-l` view. I have added a helper function (getItemDisplayData) to centralize the preparation of the data for each file/folder. Column widths for RAM and size are dynamically calculated based on the content length to ensure good alignment.

- Added script RAM usage in GB and file sizes in byte to the `ls -l` command
- Non-script files (`.lit`, `.txt`, ...) show `-` for RAM usage
- invalid script files are shown as `NaN`
- grep still works

`ls` results in the same output:
![image](https://github.com/user-attachments/assets/a62a248e-3a85-47fa-93a5-eeb8551012d7)

`ls -l` previously:
![image](https://github.com/user-attachments/assets/4458f753-2a58-47f8-9d07-69f8b30bcb18)

`ls -l` after the change:
![image](https://github.com/user-attachments/assets/4121a348-f872-4672-9cb4-0af700f4e7c0)

HansLuft778 committed:
added file size and ram usage to `ls -l`

d0sboots commented on src/Terminal/commands/ls.tsx:
```diff
@@ -109,6 +128,82 @@ export function ls(args: (string | number | boolean)[], server: BaseServer): voi
+    fileType: FileType,
+  ): { ramDisplay: string; sizeDisplay: string; nameElement: React.ReactElement } {
+    // Get name element of file
+    let nameElement: React.ReactElement;
```
By having this inside getItemDisplayData, you end up creating the react element twice: Once when you measure the sizes (and then you ignore the element) and a 2nd time when you actually use it. Because React is weird, creating elements twice like this can have correctness effects, especially when stuff like `useState` gets involved, and it always has performance implications.

Since the logic for creating the elements is actually completely separate from the sizing, I suggest you just split this into two functions.

d0sboots commented on src/Terminal/commands/ls.tsx:
```diff
@@ -180,49 +276,62 @@ export function ls(args: (string | number | boolean)[], server: BaseServer): voi
+
+    // print file based on mode
+    if (flags["-l"]) {
+      segments.forEach((segmentPath) => {
```
Optional: `forEach` has a lot of pitfalls, this can be directly replaced with `for of`. (I didn't comment on the other ones because doing one-line for-of is a little weird.)

d0sboots commented:
This looks good! The formatting was making me feel weird, but it really is how Unix `ls` works.

HansLuft778 committed:
added human readable flag to ls command

HansLuft778 committed:
updated ls command help text to include new flags

HansLuft778 commented:
I have added the `-h`  flag to show file sizes in human readable format and adjusted the `help` command text to include the changes
![image](https://github.com/user-attachments/assets/9a2eeab0-1666-4aaf-bf50-8954aaf784b5)

HansLuft778 committed:
fixed formatting

d0sboots reviewed (commented):
Dagnabbit, I never sent these pending remarks

d0sboots commented:
Oh also, if possible it's easier for us reviewers if you don't rebase your changes but merge instead. I'm gonna squash them all down to one when they're submitted anyway, and rebasing loses the history and makes it harder to review things incrementally.

HansLuft778 committed:
Address review feedback for ls -l command

HansLuft778 commented:
should be fixed now.
Regarding rebase: Usually I also prefer to merge, but I noticed the CONTRIBUTING.md mentions rebasing against dev so i chose to rebase instead of merge, sorry :D

catloversg commented on src/Terminal/commands/ls.tsx:
```diff
@@ -109,6 +130,84 @@ export function ls(args: (string | number | boolean)[], server: BaseServer): voi
+  }
+
+  // Helper function to format bytes into human-readable strings
+  function formatBytes(bytes: number, decimals = 1): string {
```
You should use `formatRam` in `src\ui\formatNumber.ts` instead of this function. We need to be consistent. Otherwise, the RAM usage printed by `ls` will not be the same as values printed everywhere else.

HansLuft778 replied:
I will use it for the script's RAM usage, but should i also for the filesize? Looks like its only compatible with GB and upwards

catloversg replied:
Regarding file size, let's wait for @d0sboots's opinion.

d0sboots replied:
We do not have a formatter that is appropriate for built-in filesizes yet. However, since you've already mostly made one, we should either scrap it (seems sad), or go all the way and finish it for general use.

To do the latter:
* Move this to ui/formatNumber as `formatFilesize`
* You're going to want to copy `formatRam` for most of the implementation. In particular, it should use `getFormatter` and you're going to need to add a similar thing to `ramSuffixList` to work with the settings switch between KB and KiB. Yup, it's a pain, but that's what you get when you sign up for human-readable units. XD

If you don't want to do all that, I'd be totally fine with just using raw numbers for filesize.

catloversg commented:
You also need to update text at:
https://github.com/bitburner-official/bitburner-src/blob/9824166dbf86b8d572fb2ca5aa4bce46ca5877fe/src/Terminal/HelpText.ts#L339

Optional: Explain the meaning of 2 new columns here. The first column is self-explanatory, but the second one is not.

d0sboots commented:
> should be fixed now. Regarding rebase: Usually I also prefer to merge, but I noticed the CONTRIBUTING.md mentions rebasing against dev so i chose to rebase instead of merge, sorry :D

That was a list of stuff to do before you (initially) submit a PR. However, I can see how it would be ambiguous. We should probably add a section about what to do and what to expect *during* the PR process to CONTRIBUTING.

HansLuft778 committed:
Update ls command help text

HansLuft778 committed:
Refactor `formatRam` to a centralized formatting functions for bytes and RAM

HansLuft778 committed:
Add tests for formatBytes function

HansLuft778 committed:
Merge branch 'dev' into ls-file-size-ram-usage

catloversg commented on src/Terminal/HelpText.ts:
```diff
@@ -336,14 +336,16 @@ export const HelpTexts: Record<string, string[]> = {
     "The 'dir' optional parameter can be used to display files/directories in another directory.",
     " ",
-    "The '-l' optional parameter allows you to force each item onto a single line.",
+    "The '-l' optional parameter allows you to force each item onto a single line, displaying two new columns for the script's RAM usage and filesize.",
```
```suggestion
    "The '-l' optional parameter allows you to force each item onto a single line, displaying two columns for the script's RAM usage and filesize.",
```
Help text/comments should be about the current state, not the history.

catloversg reviewed (commented):
I have not checked `formatNumber.ts`. I'll leave that to d0sboots.

HansLuft778 committed:
Update src/Terminal/HelpText.ts

Co-authored-by: catloversg <152669316+catloversg@users.noreply.github.com>

d0sboots commented on test/jest/ui/formatNumber.test.ts:
```diff
@@ -210,3 +213,35 @@ describe("Ram formatting", () => {
 });
+describe("Filesize formatting", () => {
+  test("With default GB mode", () => {
+    // Initial settings
```
Can you add NaN and both infs to this and the ram test? Otherwise, everything looks great!

HansLuft778 committed:
Merge branch 'dev' into ls-file-size-ram-usage

HansLuft778 committed:
add NaN and ±infinity to format number tests
